File: lib/synthesize.js
/**
 * LLM synthesis for standard Q&A mode.
 * Uses a simple, direct-answer format — no fixed multi-section template.
 */
import { getConfig, getApiKey } from './config.js';
import { resolveModel } from './chat-models.js';
import { chatCompletionsPayload } from './llm-chat-params.js';
import { splitAuthorField, joinAuthorsApaStyle } from './author-format.js';
import { sanitizeChunkTextForLlm } from './strip-running-headers.js';
import {
  isFrameworkStrictQuery,
  frameworkStrictPromptFragment,
  labelsForMatchingFrameworks,
  countChunksWithFrameworkMention,
} from './framework-queries.js';
import OpenAI from 'openai';

const SIMPLE_TEMPLATE = `You are a {identity}.

OUTPUT FORMAT (normal Q&A — not a report):
1. Start immediately with the substantive answer. Optional single heading: "## Answer" then your text; or no heading and go straight into prose/bullets.
2. Do NOT use report-style sections. Forbidden headings (do not write these at all): "Executive Summary", "Key Findings", "Evidence Summary", "Introduction", "Overview", "Discussion", "Strategic Implications", "Technology Maturity", "Adoption Barriers", or any similar multi-section report blocks.
3. If evidence is insufficient, say so in plain sentences inside the answer — still no Executive Summary / Key Findings structure.
4. Answer using ONLY the evidence excerpts. Do not invent facts or citations.

Format inside the answer only:
- Criteria / lists / steps → numbered or bullet list, each point cited where relevant
- Definitions or short factual questions → short direct prose
- Comparisons → brief analysis with inline citations

Then end with exactly one block (mandatory):

## References
One APA 7 line per source you cited (Author, A. A. (Year). Title. https://doi.org/...).

In-text citations: author surnames only — (Surname et al., Year) or (Surname & Surname, Year). Never use journal names or article titles as if they were authors. If no author: ("Short Title", n.d.).`;

function chunkToAPA7(c) {
  const parts = splitAuthorField(c.authors || '');
  const authorsStr = joinAuthorsApaStyle(parts);
  const year = (c.year || '').trim() ? ` (${(c.year || '').trim()}).` : '.';
  const title = (c.title || '').trim() || (c.source || 'Untitled');
  const doi = (c.doi || '').trim().replace(/^https?:\/\/doi\.org\//i, '');
  const url = doi ? ` https://doi.org/${doi}` : '';
  return authorsStr + year + ' ' + title + '.' + url;
}

function formatEvidence(chunks, maxExcerptChars = 8000) {
  const parts = [];
  let total = 0;
  for (let i = 0; i < chunks.length; i++) {
    const c = chunks[i];
    const text = sanitizeChunkTextForLlm(c.text, 2000);
    if (total + text.length > maxExcerptChars) break;
    total += text.length;
    const citation = chunkToAPA7(c);
    parts.push(`[Source ${i + 1}: ${citation}]\n` + text);
  }
  return parts.length ? parts.join('\n\n---\n\n') : '(No evidence provided.)';
}

export async function synthesize(query, retrievedChunks, options = {}) {
  const config = options.config || getConfig();
  const apiKey = options.api_key || getApiKey();
  if (!apiKey) throw new Error('OPENAI_API_KEY is required.');

  const llm = config.llm || {};
  // API handler already resolves; re-resolve only when synthesize is called without model (e.g. tests).
  const rawPick =
    typeof options.model === 'string' && options.model.trim() ? options.model.trim() : undefined;
  const model = resolveModel(rawPick, llm.model);
  const system_identity =
    config.system_identity ||
    'Technology Strategy Analyst for JLR specializing in emerging AEC design technologies.';

  if (!retrievedChunks || retrievedChunks.length === 0) {
    return '## References\nNone.\n\nInsufficient evidence was retrieved for this query.';
  }

  const evidence = formatEvidence(retrievedChunks);
  let systemMsg = SIMPLE_TEMPLATE.replace('{identity}', system_identity);
  if (isFrameworkStrictQuery(query)) {
    const stats = countChunksWithFrameworkMention(query, retrievedChunks);
    const fw = frameworkStrictPromptFragment(labelsForMatchingFrameworks(query));
    systemMsg = `${fw}\n\nExcerpt check: ${stats.withMention} of ${stats.total} retrieved chunks explicitly name the framework acronym or its standard full name.\n\n${systemMsg}`;
  }
  const userMsg = `Evidence:\n\n${evidence}\n\n---\n\nQuestion: ${query}`;

  const client = new OpenAI({
    apiKey,
    ...(llm.base_url ? { baseURL: llm.base_url } : {}),
  });

  const payload = chatCompletionsPayload(model, llm, [
    { role: 'system', content: systemMsg },
    { role: 'user', content: userMsg },
  ]);
  const resp = await client.chat.completions.create(payload);

  return resp.choices[0]?.message?.content ?? '';
}

File: backend/synthesize.py
"""
LLM synthesis for JLR Technology Intelligence Assistant.
Uses only retrieved evidence; supports dynamic content blocks (adaptive RAG) or legacy fixed template.
"""
from __future__ import annotations

from typing import Any

from config.load_config import load_config

try:
    from openai import OpenAI
    HAS_OPENAI = True
except ImportError:
    HAS_OPENAI = False

from backend.adaptive_rag import CONTENT_BLOCKS

LEGACY_TEMPLATE = """You are a {system_identity}

Answer the user's question using ONLY the provided evidence excerpts. Do not invent citations or add information not supported by the evidence. If evidence is insufficient, say so explicitly.

Output your response in the following structure:

## Executive Summary
(2–4 sentences.)

## Key Findings
(Bullet points grounded in the evidence. Cite the source in parentheses after each point, e.g. (Author et al., Year).)

## Adoption Barriers
(Barriers mentioned in the literature; if none found, state "No adoption barriers identified in the retrieved evidence.")

## Technology Maturity Signal
(Emerging / Growth / Mature / Decline, with brief justification from evidence.)

## Strategic Implications for JLR
(1–3 concrete, actionable implications. Prefer specific capabilities, named risk areas, or types of partnerships; avoid generic statements without detail.)

## APA References
(One line per source. APA 7 format: Author, A. A., Author, B. B., & Author, C. C. (Year). Title. Journal, Volume(Issue), pages. https://doi.org/... Do not repeat the author list or add a short form after the citation.)

Rules:
- Use only retrieved evidence. Do not invent citations.
- Cite each source you use at least once in the narrative.
- No single excerpt longer than ~500 words in your discussion.
- Do not reproduce full sections verbatim."""


def build_dynamic_prompt(block_ids: list[str], system_identity: str) -> str:
    """Build prompt from selected content block IDs. Always include sources."""
    ids = block_ids if block_ids else ["summary", "key_findings", "sources"]
    if "sources" not in ids:
        ids = list(ids) + ["sources"]
    sections = []
    for bid in ids:
        b = CONTENT_BLOCKS.get(bid)
        if b:
            sections.append(f"## {b['heading']}\n{b['instruction']}")
    body = "\n\n".join(sections)
    return f"""You are a {system_identity}

Answer the user's question using ONLY the provided evidence excerpts. Do not invent citations or add information not supported by the evidence. If evidence is insufficient for a section, say so briefly or omit that section.

Output your response in the following structure (include only these sections):

{body}

Rules:
- Use only retrieved evidence. Do not invent citations.
- Cite each source you use at least once (e.g. (Author et al., Year)).
- When multiple sources support or contrast a point, cite more than one.
- For Strategic Implications: be specific (e.g. which capability, which risk); avoid generic advice.
- No single excerpt longer than ~500 words in your discussion.
- Do not reproduce full sections verbatim."""


def chunk_to_apa7(c: dict[str, Any]) -> str:
    """Build APA 7 citation from chunk fields (avoids duplicated/stored malformed apa_citation)."""
    import re
    raw = (c.get("authors") or "").strip()
    parts = [p.strip() for p in re.split(r"[;,]+", raw) if p.strip()] if raw else []
    if not parts:
        authors_str = "Unknown"
    elif len(parts) == 1:
        authors_str = parts[0]
    else:
        authors_str = ", ".join(parts[:-1]) + ", & " + parts[-1]
    y = (c.get("year") or "").strip()
    year = f" ({y})." if y else "."
    title = (c.get("title") or "").strip() or (c.get("source") or "Untitled")
    doi = (c.get("doi") or "").strip()
    if doi and doi.lower().startswith("http"):
        doi = re.sub(r"^https?://doi\.org/", "", doi, flags=re.I)
    url = " https://doi.org/" + doi if doi else ""
    return authors_str + year + " " + title + "." + url


def get_openai_client(api_key: str | None = None, base_url: str | None = None) -> "OpenAI":
    if not HAS_OPENAI:
        raise RuntimeError("openai is required. Install with: pip install openai")
    kwargs: dict = {"api_key": api_key}
    if base_url:
        kwargs["base_url"] = base_url
    return OpenAI(**kwargs)


def format_evidence(chunks: list[dict[str, Any]], max_excerpt_chars: int = 8000) -> str:
    """Format retrieved chunks for the prompt. Use clean APA 7 citation per chunk."""
    parts = []
    total = 0
    for i, c in enumerate(chunks):
        text = (c.get("text") or "")[:2000]
        if total + len(text) > max_excerpt_chars:
            break
        total += len(text)
        citation = chunk_to_apa7(c)
        header = f"[Source {i+1}: {citation}]\n"
        parts.append(header + text)
    return "\n\n---\n\n".join(parts) if parts else "(No evidence provided.)"


def synthesize(
    query: str,
    retrieved_chunks: list[dict[str, Any]],
    config: dict[str, Any] | None = None,
    *,
    api_key: str | None = None,
    blocks: list[str] | None = None,
) -> str:
    """
    Generate structured analytical response from query and retrieved chunks.
    If blocks is provided, uses dynamic prompt with only those sections; else legacy template.
    """
    if not HAS_OPENAI:
        raise RuntimeError("openai is required. Install with: pip install openai")
    config = config or load_config()
    llm_cfg = config.get("llm", {})
    model = llm_cfg.get("model", "gpt-4o")
    temperature = llm_cfg.get("temperature", 0.3)
    max_tokens = llm_cfg.get("max_tokens", 2048)
    system_identity = config.get("system_identity", "Technology Strategy Analyst for JLR specializing in emerging AEC design technologies.")

    if not retrieved_chunks:
        return (
            "## Executive Summary\n"
            "Insufficient evidence was retrieved for this query. No analytical response can be generated without source material.\n\n"
            "## APA References\n"
            "None."
        )

    evidence = format_evidence(retrieved_chunks)
    if blocks:
        system_msg = build_dynamic_prompt(blocks, system_identity)
    else:
        system_msg = LEGACY_TEMPLATE.format(system_identity=system_identity)
    user_msg = f"Evidence:\n\n{evidence}\n\n---\n\nUser question: {query}"

    base_url = llm_cfg.get("base_url") or None
    client = get_openai_client(api_key=api_key, base_url=base_url)
    resp = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system_msg},
            {"role": "user", "content": user_msg},
        ],
        temperature=temperature,
        max_tokens=max_tokens,
    )
    return resp.choices[0].message.content or ""


if __name__ == "__main__":
    import os
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
    from backend.retrieve import retrieve
    from config.load_config import get_api_key

    config = load_config()
    chunks = retrieve("What are adoption barriers for BIM in construction?", config=config, api_key=get_api_key())
    answer = synthesize("What are adoption barriers for BIM in construction?", chunks, config=config, api_key=get_api_key())
    print(answer)
